src/main/java/com/palantir/docker/compose/DockerComposeRule.java

```diff
@@ -102,7 +121,20 @@ protected LogCollector logCollector() {
     public void before() throws IOException, InterruptedException {
         log.debug("Starting docker-compose cluster");
         dockerCompose().build();
-        dockerCompose().up();
+
+        try {
+            dockerCompose().up();
+        } catch (DockerExecutionException e) {
+            Set<String> conflictingContainerNames = getConflictingContainerNames(e.getMessage());
```

Comment:
I went back and forth on this, and played around with returning a result from `dockerCompose.up()`. However, I didn't really like the result -- basically, you would end up in a place where `dockerCompose.up()` could return a value, and the content of the value would need to be inspected to determine whether or not the operation was successful. This just didn't feel right to me, especially if the rest of the contracts weren't being altered.

With the current structure, I think that, if `dockerCompose.up()` returns normally, it should mean that the operation succeeded and that the user can assume that the compose cluster is now "up". Any case besides that should be exceptional.

I've cleaned this up by creating a separate implementation of `DockerComposeRule` that encodes this logic. Its implementation of `up()` delegates to the internal one, catches the exceptional state it supports and then attempts to re-try. This centralizes this logic in a single place, but still keeps the over-all contract -- if the wrapped `up()` returns, it means that the cluster did in fact come up, while if it couldn't recover, it will throw an exception. The logic no longer leaks out to `DockerComposeRule`, and I think it's consistent with the way other things in this library are implemented.